README.md
[![Open Source Love png1](https://badges.frapsoft.com/os/v1/open-source.png?v=103)](https://github.com/ellerbrock/open-source-badges/)
[![GPLv3 license](https://img.shields.io/badge/License-GPLv3-blue.svg)](http://perso.crans.org/besson/LICENSE.html)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square)](http://makeapullrequest.com)
[![Say Thanks!](https://img.shields.io/badge/Say%20Thanks-!-1EAEDB.svg)](https://saythanks.io/to/daksh7011)

# Eagle Crypto Suite

:star: Star this repository — it helps!

[Eagle](https://daksh7011.com/eagle) is an cryptographic suite  that bundles various encryption and hashing methods 
without database backend and minimal codebase.

[![forthebadge](https://forthebadge.com/images/badges/check-it-out.svg)](https://daksh7011.com/eagle)

# Table of contents
* [Requirements](#requirements)
* [Features](#features)
* [Installation](#installation)
* [Contribution Guide](#contribution-guide)
* [License](#license)
* [Links](#links)

# Requirements
* PHP => 7.1
* Apache [mod_rewrite](https://httpd.apache.org/docs/current/mod/mod_rewrite.html) extension on web server

# Features
* Binds hashing and string related algorithms with less than 200 lines of core class code.
* Only one class to handle everything in project
* Clean code with comments about almost everything
* Useful for encryption, decryption, encoding, decoding and playing with different hashes

# Installation
1. Clone the repository on your web-server 
2. Modify the name of your website in [configuration](/includes/config.php) file and you are good to go. 
``` php
$website['name'] = '<your awsome project name>';
```


# Contribution Guide
Please take a look at the [contributing](CONTRIBUTING.md) guidelines if you're interested in helping!

# License
The Eagle Project is licensed under the terms of the GPL Open Source license version 3 and is available for free.
You should have received a copy of the GNU General Public License along with this program. 
If not, see [GPL-v3](https://www.gnu.org/licenses/gpl-3.0.html)

# Links
[Website](https://gitlab.com/daksh7011/eagle)

[Issue Tracker](https://gitlab.com/daksh7011/eagle/issues)

[Say Thanks](https://saythanks.io/to/daksh7011) - It motivates me.

**Thanks to**
 
* [Duck Duck Go](https://duckduckgo.com/) for not tracking me while trying to search what went wrong with my code.

* [Stack Overflow](https://stackoverflow.com/) for providing such platform to help fellow developers.

* [Siddhi Agrawal](https://gitlab.com/echo-siddhi) for continuous assistance when I hit my head on the wall.

[![forthebadge](https://forthebadge.com/images/badges/powered-by-oxygen.svg)](https://en.wikipedia.org/wiki/Oxygen)

CHANGELOG
v 0.1 
    -initialized project
v 0.2
    -initialized bootstrap files
    -added server rewrites
v 0.3
    -added project wide config file that handles name of project and misc things
    -added magical class which will be the core of Project Eagle. this class is responsible for setting website
        protocol, url and fetch current url on the fly. more to come on same class.

V 1.0
    file: eagle.class.php
        -added methods to prepare slugs, names, algorithms.
        -added method to exploit available string algorithms in php core (added base64 encode and base64 decode)
        -added method to count available hash and string methods to show it in main page hero.
        -added method to convert algorithm names to appropriate slugs
        -added magic method to use algorithms.

    -Added working front end with custom css style based on bootstrap framework
V 1.1
    -Hot-fix: updated method name in eagle.class.php
    -modified readme with relevant info
    -modified contribution guide.
V 1.2
    -refactor the method names with camelCase and class name with PascalCase
    -modified bugged contribution guide with correct guide

index.php
<?php
/**
 *
 *  * Open Source Project made by Daksh (daksh7011.com)
 *  * Do NOT remove this excerpt even if you fork this or use it by any means.
 *  * This project is Licenced under GNU GENERAL PUBLIC LICENSE Version 3
 *
 *
 */

// require the config file.
require_once('includes/config.php');

// set the title & description.
$type = (isset($_GET['type']) && in_array($_GET['type'], array('string', 'hash')) ? $_GET['type'] : false);
if ($type) {
    $website['title'] = ucfirst($type) . ' Algorithms';
    $website['description'] = 'Use our ' . $type . ' algorithms on your strings.';
} else {
    $website['title'] = 'Home';
    $website['description'] = $website['name'] . ' is a free online service providing ' . $website['count'] . ' hash and string algorithms.';
}
?>
<!DOCTYPE html >
<html lang="en">
<head>

    <!-- meta tags -->
    <meta http-equiv="content-type" content="text/html; charset=utf-8"/>
    <title><?php echo $website['title']; ?></title>
    <meta name="description" content="<?php echo $website['description']; ?>"/>

    <!-- favicon that doesnt exist yet -->
    <link href="assets/img/favicon.ico" rel="shortcut icon" type="image/x-icon"/>
    <link href="assets/img/favicon.ico" rel="icon" type="image/x-icon"/>

    <!-- stylesheets -->
    <link rel="stylesheet"
          href="https://use.fontawesome.com/releases/v5.5.0/css/all.css"
          integrity="sha384-B4dIYHKNBt8Bc12p+WXckhzcICo0wtJAoU8YZTY5qE0Id1GSseTk6S+L3BlXeVIU"
          crossorigin="anonymous">
    <link href="assets/css/bootstrap.min.css" rel="stylesheet" type="text/css"/>
    <link href="assets/css/styles.css?t=<?php echo time(); ?>" rel="stylesheet"
          type="text/css"/>
    <link href="https://fonts.googleapis.com/css?family=Lato:400,300,700" rel="stylesheet" type="text/css">
    <!-- bootstrap javascript -->
    <script src="assets/js/bootstrap.js" type="text/javascript"></script>

</head>
<body>

<!-- include the website header and hero  -->
<?php include_once('includes/html/header.php'); ?>
<?php include_once('includes/html/hero.php'); ?>

<!-- website main -->
<div id="main">
    <?php if (!$type || ($type && $type == 'string')): ?>
        <div id="string" class="section">
            <div class="container">
                <div class="section-information">
                    <h2>String algorithms</h2>
                </div>
                <div class="row">
                    <?php foreach ($website['algorithms'] as $algorithm): ?>
                        <?php if ($algorithm['type'] == 'string'): ?>
                            <div class="col-md-4">
                                <a href="<?php echo $algorithm['url']; ?>" title="<?php echo $algorithm['name']; ?>">
                                    <div class="panel panel-default panel-tool">
                                        <div class="panel-body">
                                            <!-- Dump string algorithms-->
                                            <?php echo $algorithm['name']; ?>

                                        </div>
                                    </div>
                                </a>
                            </div>
                        <?php endif; ?>
                    <?php endforeach; ?>
                </div>
            </div>
        </div>
    <?php endif; ?>


</div>

<!--include the website footer file.-->
<?php include_once('includes/html/footer.php'); ?>

</body>
</html>

includes/Eagle.class.php
<?php
/**
 *
 *  * Open Source Project made by Daksh (daksh7011.com)
 *  * Do NOT remove this excerpt even if you fork this or use it by any means.
 *  * This project is Licenced under GNU GENERAL PUBLIC LICENSE Version 3
 *
 *
 */

class Eagle
{

    //get protocol whether it is secured or plain
    public static function websiteProtocol()
    {
        return (isset($_SERVER['HTTPS']) && (!empty($_SERVER['HTTPS']) || $_SERVER['HTTPS'] !== 'off') ? 'https' : 'http');
    }

    //get website url from global var and concat it with protocol
    public static function websiteUrl()
    {
        //$dir = dirname($_SERVER['PHP_SELF']); //doesnt work
        //$dir = dirname($_SERVER['REQUEST_URI']); //doesnt work
        //$dir = dirname($_SERVER['SCRIPT_FILENAME']); //doesnt work either
        return self::websiteProtocol() . '://' . $_SERVER['HTTP_HOST'] . dirname($_SERVER['PHP_SELF']) . '/';
    }

    //get current full request path including the querystring.
    public static function websiteCurrent()
    {
        return self::websiteProtocol() . '://' . $_SERVER['HTTP_HOST'] . $_SERVER['REQUEST_URI'];
    }

    //prepare slugs
    public static function algorithmSlug($string)
    {
        $slug = strtolower($string);
        $slug = preg_replace('/[^a-z0-9]+/', '-', $slug);
        return trim($slug, '-');
    }

    //prepare name of algorithm
    public static function algorithmName($string)
    {
        $name = strtolower($string);
        $name = preg_replace('/[^a-z0-9\.\,\s]+/', ' ', $name);
        return trim($name, ' ');
    }

    //available string algorithms in core php
    public static function stringAlgorithms()
    {
        $functions = array(
            "base64_encode",
            "base64_decode"
        );

        //store slugs and url scheme for string crypto algorithms
        $algorithms = array();
        foreach ($functions as $function) {
            if (function_exists($function)) {
                $algorithm = array();
                $algorithm['slug'] = self::algorithmSlug($function);
                $algorithm['name'] = self::algorithmName($function);
                $algorithm['url'] = self::websiteUrl() . 'algorithm/' . $algorithm['slug'];
                $algorithm['algorithm'] = $function;
                $algorithm['type'] = 'string';
                $algorithms[] = $algorithm;
            }
        }
        return (!empty($algorithms) ? $algorithms : false);
    }

    //hashing algorithms yet to implement

    //count available algorithms
    public static function allAlgorithms()
    {
        $hash = null;
        $string = self::stringAlgorithms();
        return ($hash && $string ? array_merge($hash, $string) : ($hash ? $hash : $string));
    }

    //convert algorithm names to appropriate slugs
    public static function slugToAlgorithm($slug)
    {
        $algorithms = self::allAlgorithms();
        foreach ($algorithms as $algorithm) {
            if ($algorithm['slug'] == $slug) {
                return $algorithm;
                break;
            }
        }
        return false;
    }


    //the main magic happens here
    public static function magicMethod($algorithm, $opts)
    {
        //validate if string is not null
        if (!isset($opts['string']) || empty($opts['string'])) {
            return 'the string input is empty, please provide a string.';
        } //check algorithm type
        else if ($algorithm['type'] == 'string') {
            return $algorithm['algorithm']($opts['string']);
        } //when custom salt is available
        else if (isset($opts['string']) && isset($opts['salt'])) {
            return hash($algorithm['algorithm'], $opts['string'] . $opts['salt']);
        } //when custom salt doesnt exist
        else {
            return hash($algorithm['algorithm'], $opts['string']);
        }
    }

}

?>

includes/eagle.class.php
<?php
/**
 *
 *  * Open Source Project made by Daksh (daksh7011.com)
 *  * Do NOT remove this excerpt even if you fork this or use it by any means.
 *  * This project is Licenced under GNU GENERAL PUBLIC LICENSE Version 3
 *
 *
 */

class eagle
{

    //get protocol whether it is secured or plain
    public static function website_protocol()
    {
        return (isset($_SERVER['HTTPS']) && (!empty($_SERVER['HTTPS']) || $_SERVER['HTTPS'] !== 'off') ? 'https' : 'http');
    }

    //get website url from global var and concat it with protocol
    public static function website_url()
    {
        return self::website_protocol() . '://' . $_SERVER['HTTP_HOST'] . dirname($_SERVER['PHP_SELF']) . '/';
    }

    //get current url
    public static function website_current()
    {
        return self::website_protocol() . '://' . $_SERVER['HTTP_HOST'] . $_SERVER['REQUEST_URI'];
    }

    //prepare slugs
    public static function algorithm_slug($string)
    {
        $slug = strtolower($string);
        $slug = preg_replace('/[^a-z0-9]+/', '-', $slug);
        return trim($slug, '-');
    }

    //prepare name of algorithm
    public static function algorithm_name($string)
    {
        $name = strtolower($string);
        $name = preg_replace('/[^a-z0-9\.\,\s]+/', ' ', $name);
        return trim($name, ' ');
    }

    //available string algorithms in core php
    public static function string_algorithms()
    {
        $functions = array(
            "base64_encode",
            "base64_decode"
        );

        //store slugs and url scheme for string crypto algorithms
        $algorithms = array();
        foreach ($functions as $function) {
            if (function_exists($function)) {
                $algorithm = array();
                $algorithm['slug'] = self::algorithm_slug($function);
                $algorithm['name'] = self::algorithm_name($function);
                $algorithm['url'] = self::website_url() . 'algorithm/' . $algorithm['slug'];
                $algorithm['algorithm'] = $function;
                $algorithm['type'] = 'string';
                $algorithms[] = $algorithm;
            }
        }
        return (!empty($algorithms) ? $algorithms : false);
    }

    //hashing algorithms yet to implement

    //count available algorithms
    public static function all_algorithms()
    {
        $hash = null;
        $string = self::string_algorithms();
        return ($hash && $string ? array_merge($hash, $string) : ($hash ? $hash : $string));
    }

    //convert algorithm names to appropriate slugs
    public static function slug_to_algorithm($slug)
    {
        $algorithms = self::all_algorithms();
        foreach ($algorithms as $algorithm) {
            if ($algorithm['slug'] == $slug) {
                return $algorithm;
                break;
            }
        }
        return false;
    }


    //the main magic happens here
    public static function magic_method($algorithm, $opts)
    {
        //validate if string is not null
        if (!isset($opts['string']) || empty($opts['string'])) {
            return 'the string input is empty, please provide a string.';
        } //check algorithm type
        else if ($algorithm['type'] == 'string') {
            return $algorithm['algorithm']($opts['string']);
        } //when custom salt is available
        else if (isset($opts['string']) && isset($opts['salt'])) {
            return hash($algorithm['algorithm'], $opts['string'] . $opts['salt']);
        } //when custom salt doesnt exist
        else {
            return hash($algorithm['algorithm'], $opts['string']);
        }
    }

}

?>

includes/html/footer.php
<?php
/**
 *
 *  * Open Source Project made by Daksh (daksh7011.com)
 *  * Do NOT remove this excerpt even if you fork this or use it by any means.
 *  * This project is Licenced under GNU GENERAL PUBLIC LICENSE Version 3
 *
 *
 */

?>
<!-- copyright bar -->
<div id="footer-copyright">
   <div class="container">
         <span class="pull-left"><?php echo $website['name'];?>, Proudly Open-Source project.</span>
   </div>
</div>
<!-- website footer -->
<footer>
   <div class="container">

      <ul class="pull-left">
         <li><a href="<?php echo $website['url'];?>">home</a></li>
         <li><span> | </span></li>
         <li><a href="<?php echo $website['url'];?>string-algorithms">string algorithms</a></li>
         <li><span> | </span></li>
         <li><a href="<?php echo $website['url'];?>hash-algorithms">hash algorithms</a></li>
      </ul>

      <ul class="pull-right">
         <li><a href="https://twitter.com/SlothieSmooth" target="_blank"><i class="fab fa-twitter"></i></a></li>
      </ul>

   </div>
</footer>
<script src="https://code.jquery.com/jquery-3.3.1.min.js"
        integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8="
        crossorigin="anonymous"></script>
<script src="<?php echo $website['url'];?>assets/js/bootstrap.min.js"></script>

includes/html/header.php
<?php
/**
 *
 *  * Open Source Project made by Daksh (daksh7011.com)
 *  * Do NOT remove this excerpt even if you fork this or use it by any means.
 *  * This project is Licenced under GNU GENERAL PUBLIC LICENSE Version 3
 *
 *
 */

?>
<nav class="navbar navbar-default">
    <div class="container">
        <div class="navbar-header">
            <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar"
                    aria-expanded="false" aria-controls="navbar">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            <a class="navbar-brand" href="<?php echo $website['url']; ?>">
                <img src="<?php echo $website['url']; ?>/assets/img/logo.png"
                     alt="<?php echo $website['name']; ?>"/></a>
        </div>
        <div id="navbar" class="collapse navbar-collapse">
            <ul class="nav navbar-nav">
                <li <?php if ((isset($type) && $type == 'string') || (isset($algorithm['type']) &&
                        $algorithm['type'] == 'string')) {
                    echo 'class="active"';
                } ?>>
                    <a href="<?php echo $website['url']; ?>string-algorithms">String algorithms</a>
                </li>
                <li <?php if ((isset($type) && $type == 'hash') || (isset($algorithm['type']) &&
                        $algorithm['type'] == 'hash')) {
                    echo 'class="active"';
                } ?>>
                    <a href="<?php echo $website['url']; ?>hash-algorithms">Hash algorithms</a>
                </li>
                <li>
                    <a href="gitlab.com/daksh7011/eagle" target="_blank">Fork this</a>
                </li>
            </ul>
        </div>
    </div>
</nav>

algorithm.php
<?php
/**
 *
 *  * Open Source Project made by Daksh (daksh7011.com)
 *  * Do NOT remove this excerpt even if you fork this or use it by any means.
 *  * This project is Licenced under GNU GENERAL PUBLIC LICENSE Version 3
 *
 *
 */

// require the website config file.
require_once('includes/config.php');

// set the title, description & response.
$algorithm = (isset($_GET['slug']) ? Eagle::slugToAlgorithm($_GET['slug']) : false);
if ($algorithm) {
    $website['title'] = ucwords($algorithm['name']);
    $website['description'] = 'Use the ' . $algorithm['name'] . ' ' . $algorithm['type'] . ' algorithm on your strings.';
    $website['response'] = (isset($_POST['submit']) ? Eagle::magicMethod($algorithm, $_POST) : false);
} else {
    header('location: ../');
    exit;
}
?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
        "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html lang="en"">
<head>

    <!-- meta tags -->
    <meta http-equiv="content-type" content="text/html; charset=utf-8"/>
    <title><?php echo $website['title']; ?></title>
    <meta name="description" content="<?php echo $website['description']; ?>"/>

    <!-- favicon that doesnt exist -->
    <link href="<?php echo $website['url']; ?>assets/img/favicon.ico" rel="shortcut icon" type="image/x-icon"/>
    <link href="<?php echo $website['url']; ?>assets/img/favicon.ico" rel="icon" type="image/x-icon"/>

    <!-- stylesheets -->
    <link href="<?php echo $website['url']; ?>assets/css/font-awesome.min.css" rel="stylesheet" type="text/css"/>
    <link href="<?php echo $website['url']; ?>assets/css/bootstrap.min.css" rel="stylesheet" type="text/css"/>
    <link href="<?php echo $website['url']; ?>assets/css/styles.css?t=<?php echo time(); ?>" rel="stylesheet"
          type="text/css"/>
    <link href="https://fonts.googleapis.com/css?family=Lato:400,300,700" rel="stylesheet" type="text/css">
    <!-- javascript -->
    <script src="<?php echo $website['url']; ?>assets/js/bootstrap.js" type="text/javascript"></script>

</head>
<body>

<!--include the header and hero files.-->
<?php include_once('includes/html/header.php'); ?>
<?php include_once('includes/html/hero.php'); ?>

<!-- main -->
<div id="main" class="algo">
    <div class="section">
        <div class="container">
            <div class="section-information">
                <h2><?php echo ucfirst($algorithm['name']); ?></h2>
                <p>Use the <?php echo $algorithm['name']; ?> <?php echo $algorithm['type']; ?> algorithm on your
                    strings.</p>
            </div>
            <form method="post" action="<?php echo $website['current']; ?>" id="algorithm-form">
                <?php if ($algorithm['type'] == 'hash'): ?>
                    <div class="form-group"><input type="text" name="salt" placeholder="salt (optional)"
                                                   class="form-control input-lg"/></div>
                <?php endif; ?>
                <div class="form-group"><textarea name="string" placeholder="string"
                                                  class="form-control input-lg"><?php if ($website['response']) {
                            echo $website['response'];
                        } ?></textarea></div>
                <button type="submit" name="submit" class="btn btn-blue btn-lg"><?php echo $algorithm['name']; ?> my
                    string
                </button>
            </form>
        </div>
    </div>
    <div class="section">
        <div class="container">
            <div class="section-information">
                <h2>More algorithms</h2>
                <p>Here is some more algorithms for you to use on your strings.</p>
            </div>
            <div class="algorithm-list">
                <?php foreach ($website['more'] as $key): ?>
                    <div class="col-md-3">
                        <a href="<?php echo $website['algorithms'][$key]['url']; ?>"
                           title="<?php echo $website['algorithms'][$key]['name']; ?>">
                            <div class="panel panel-default panel-tool">
                                <div class="panel-body">
                                    <?php echo $website['algorithms'][$key]['name']; ?>
                                </div>
                            </div>
                        </a>
                    </div>
                <?php endforeach; ?>
            </div>
        </div>
    </div>
</div>

<!-- include the footer file. -->
<?php include_once('includes/html/footer.php'); ?>

</body>
</html>